streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java

```diff
@@ -31,35 +31,90 @@
 import java.util.Collections;
 import java.util.Deque;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.LinkedList;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.TimeUnit;
 
-import static org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsConventions.threadLevelSensorName;
-
 public class StreamsMetricsImpl implements StreamsMetrics {
     private final Metrics metrics;
     private final Map<String, String> tags;
     private final Map<Sensor, Sensor> parentSensors;
-    private final Deque<String> ownedSensors = new LinkedList<>();
     private final Sensor skippedRecordsSensor;
+    private final String threadName;
+
+    private final Deque<String> threadLevelSensors = new LinkedList<>();
+    private final Map<String, Deque<String>> cacheLevelSensors = new HashMap<>();
 
     public StreamsMetricsImpl(final Metrics metrics, final String threadName) {
         Objects.requireNonNull(metrics, "Metrics cannot be null");
+        this.threadName = threadName;
 
         this.metrics = metrics;
-        this.tags = StreamsMetricsConventions.threadLevelTags(threadName, Collections.<String, String>emptyMap());
+
+        final HashMap<String, String> tags = new LinkedHashMap<>();
+        tags.put("client-id", threadName);
+        this.tags = Collections.unmodifiableMap(tags);
+
         this.parentSensors = new HashMap<>();
 
-        skippedRecordsSensor = metrics.sensor(threadLevelSensorName(threadName, "skipped-records"), Sensor.RecordingLevel.INFO);
+        skippedRecordsSensor = threadLevelSensor("skipped-records", Sensor.RecordingLevel.INFO);
         skippedRecordsSensor.add(metrics.metricName("skipped-records-rate", "stream-metrics", "The average per-second number of skipped records", tags), new Rate(TimeUnit.SECONDS, new Count()));
         skippedRecordsSensor.add(metrics.metricName("skipped-records-total", "stream-metrics", "The total number of skipped records", tags), new Total());
-        ownedSensors.push(skippedRecordsSensor.name());
     }
 
-    public final Metrics registry() {
-        return metrics;
+    public final Sensor threadLevelSensor(final String sensorName,
+                                          final Sensor.RecordingLevel recordingLevel,
+                                          final Sensor... parents) {
+        synchronized (threadLevelSensors) {
+            final String fullSensorName = threadName + "." + sensorName;
```

Comment:
I started to do it that way, but I realized that right now StreamsMetricsImpl is only created and used in the thread level (its constructor takes a thread name), so that map would only ever have one key. It seemed needless to introduce that extra machinery for thread-level metrics if it's not necessary, and in fact it would be misleading, since it would suggest that the SMI was somehow managing sensors for more than one thread.

I figured that if, in the future, we wanted to make SMI a singleton for the whole app, we could easily retrofit threadLevelSensor.

It might be better as a thread-level class, though, so we don't have to worry about contention.

Reply:
Makes sense, we can do that later.